File: investimentos-compartilhar/assets/js/modelo/projecao.js
/* Projeção mês a mês, em moeda de hoje.
 *
 * 1. Tudo em termos reais: R$ 1 milhão em 2046 não é R$ 1 milhão.
 * 2. O que é contratado (renda fixa) separado do que é premissa.
 * 3. Proventos vêm do DY observado, não de chute — somar um retorno
 *    total "esperado" ao dividendo contaria o dinheiro duas vezes. */
(function () {
  const FC = window.FC;
  const MESES = 12;
  const mensal = (a) => Math.pow(1 + a / 100, 1 / MESES) - 1;
  const real = (nom, ipca) => ((1 + nom / 100) / (1 + ipca / 100) - 1) * 100;

  function perfilDosPilares(dados, macroLp, ipca) {
    const pilares = {};
    for (const a of dados.ativos) {
      if (!a.posicao || !a.pilar) continue;
      const dyM = (a.metricas || []).find((m) => m.chave === "dy" && m.valor);
      const dy = dyM ? dyM.valor : 0;
      const p = (pilares[a.pilar] = pilares[a.pilar] || { valor: 0, dyp: 0, titulos: [] });
      p.valor += a.posicao.atual;
      p.dyp += a.posicao.atual * dy;
    }
    for (const p of Object.values(pilares)) p.dy = p.valor ? p.dyp / p.valor : 0;

    const caixa = (pilares.caixa = pilares.caixa || { valor: 0, dy: 0, titulos: [] });
    const cdi = macroLp.cdi;
    for (const r of dados.rendaFixa) {
      const aplic = r.valor_aplicado || 0;
      if (!aplic) continue;
      let nominal;
      if (r.taxa == null) nominal = cdi;
      else if (r.tipo === "cdi") nominal = (cdi * r.taxa) / 100;
      else if (r.tipo === "ipca") nominal = ((1 + ipca / 100) * (1 + r.taxa / 100) - 1) * 100;
      else nominal = r.taxa;
      caixa.valor += aplic;
      caixa.titulos.push({ nome: r.nome, valor: aplic, real: real(nominal, ipca), nominal });
    }
    const tt = FC.soma(caixa.titulos, (t) => t.valor);
    caixa.taxa_real = tt ? FC.soma(caixa.titulos, (t) => t.valor * t.real) / tt : 0;

    if (dados.agro && dados.agro.valorRebanho > 0) pilares.agro = { valor: dados.agro.valorRebanho, dy: 0, titulos: [] };
    return pilares;
  }

  function destino(saldos, alvos, modo) {
    const ks = Object.keys(saldos);
    const total = ks.reduce((s, k) => s + saldos[k], 0);
    const somaAlvos = ks.reduce((s, k) => s + (Number(alvos[k]) || 0), 0);
    const proporcional = () => Object.fromEntries(ks.map((k) => [k, (Number(alvos[k]) || 0) / somaAlvos]));
    if (!somaAlvos) return Object.fromEntries(ks.map((k) => [k, 1 / ks.length]));
    if (!total) return proporcional();
    if (modo === "proporcional") return proporcional();
    const faltas = Object.fromEntries(ks.map((k) => [k, Math.max(0, (total * (Number(alvos[k]) || 0)) / 100 - saldos[k])]));
    const sf = Object.values(faltas).reduce((a, b) => a + b, 0);
    if (sf <= 0) return proporcional();
    return Object.fromEntries(ks.map((k) => [k, faltas[k] / sf]));
  }

  function simula(pilares, prem, alvos, ajuste) {
    const meses = Math.round(Number(prem.horizonte_anos) || 20) * MESES;
    const aporte = Number(prem.aporte_mensal) || 0;
    const reinveste = !!prem.reinvestir_proventos;
    const modo = prem.distribuicao_aporte || "rebalancear";
    const val = prem.valorizacao_real_anual || {};
    const saldos = Object.fromEntries(Object.entries(pilares).map(([k, p]) => [k, p.valor]));
    const dy = Object.fromEntries(Object.entries(pilares).map(([k, p]) => [k, p.dy || 0]));
    // se não há nenhum pilar (carteira vazia), o aporte cria os pilares da meta
    if (!Object.keys(saldos).length) for (const k of Object.keys(alvos)) if (Number(alvos[k]) > 0) { saldos[k] = 0; dy[k] = 0; }
    const fator = Object.fromEntries(Object.keys(saldos).map((k) => [k,
      k === "caixa" ? mensal((pilares.caixa || {}).taxa_real || 0) : mensal((Number(val[k]) || 0) + ajuste)]));

    const pat = [], renda = [];
    let aportado = 0, prov = 0;
    const inicial = Object.values(saldos).reduce((a, b) => a + b, 0);
    for (let m = 1; m <= meses; m++) {
      for (const k in saldos) saldos[k] *= 1 + fator[k];
      const rendaMes = Object.keys(saldos).filter((k) => k !== "caixa").reduce((s, k) => s + (saldos[k] * dy[k]) / 100 / MESES, 0);
      prov += rendaMes;
      const entrada = aporte + (reinveste ? rendaMes : 0);
      aportado += aporte;
      const pesos = destino(saldos, alvos, modo);
      for (const [k, w] of Object.entries(pesos)) saldos[k] += entrada * w;
      if (m % 3 === 0 || m === meses) {
        pat.push([m / MESES, Object.values(saldos).reduce((a, b) => a + b, 0)]);
        renda.push([m / MESES, rendaMes]);
      }
    }
    const final = Object.values(saldos).reduce((a, b) => a + b, 0);
    return {
      patrimonio: pat, renda, final, inicial, aportado, proventos: prov,
      valorizacao: final - inicial - aportado - (reinveste ? prov : 0),
      renda_mensal_final: renda.length ? renda.at(-1)[1] : 0,
      renda_mensal_hoje: Object.keys(pilares).filter((k) => k !== "caixa").reduce((s, k) => s + (pilares[k].valor * (dy[k] || 0)) / 100 / MESES, 0),
    };
  }

  function projeta(dados, prem) {
    const macro = dados.macro || {};
    const lp = prem.macro_longo_prazo || {};
    const cdi = lp.cdi != null ? lp.cdi : macro.cdi || 10;
    const ipca = lp.ipca != null ? lp.ipca : macro.ipca_12m || 4;
    const macroLp = { cdi, ipca, juro_real: real(cdi, ipca) };
    const alvos = dados.alocacao_alvo || {};
    const pilares = perfilDosPilares(dados, macroLp, ipca);
    const cen = prem.cenarios || {};
    const base = simula(pilares, prem, alvos, 0);
    const pess = simula(pilares, prem, alvos, Number(cen.pessimista ?? -3));
    const otim = simula(pilares, prem, alvos, Number(cen.otimista ?? 3));
    const val = prem.valorizacao_real_anual || {};
    const ordem = FC.PILARES.map((p) => p.chave);
    const resumo = Object.entries(pilares).map(([k, p]) => k === "caixa"
      ? { chave: k, nome: FC.PILAR_NOME[k], valor: p.valor, dy: null, valorizacao: null, total: p.taxa_real || 0, titulos: p.titulos }
      : { chave: k, nome: FC.PILAR_NOME[k], valor: p.valor, dy: p.dy, valorizacao: Number(val[k]) || 0, total: (Number(val[k]) || 0) + p.dy, titulos: [] })
      .sort((a, b) => ordem.indexOf(a.chave) - ordem.indexOf(b.chave));
    const caixa = resumo.find((l) => l.chave === "caixa");
    const rv = resumo.filter((l) => l.chave !== "caixa" && l.valor > 0).map((l) => l.total);
    return {
      base, pessimista: pess, otimista: otim, pilares: resumo, macro_lp: macroLp,
      usando_macro_de_hoje: lp.cdi == null, premissas: prem,
      alerta_caixa: !!caixa && caixa.valor > 0 && rv.length > 0 && caixa.total > Math.max(...rv),
      titulos_caixa: caixa ? caixa.titulos : [], caixa_real: caixa ? caixa.total : 0,
    };
  }

  FC.projecao = { projeta };
})();

File: investimentos-compartilhar/legado-python/src/modelo/projecao.py
"""Projeção da carteira mês a mês, em moeda de hoje.

Três princípios, porque é aqui que ferramenta financeira mais engana:

1. TUDO EM TERMOS REAIS. Projetar nominal produz números grandes e
   sem significado — R$ 1 milhão em 2046 não é R$ 1 milhão. Aqui o
   patrimônio projetado está sempre no poder de compra de hoje.

2. O DETERMINÍSTICO SEPARADO DA PREMISSA. Um CDB a 103,7% do CDI tem
   retorno conhecido dado o cenário de juros; a valorização de uma ação
   não tem. As duas coisas nunca aparecem com o mesmo peso de certeza.

3. PROVENTOS VÊM DO DADO, NÃO DE CHUTE. O rendimento projetado usa o
   dividend yield que cada ativo efetivamente pagou nos últimos 12
   meses, e a valorização entra à parte — somar um retorno total
   "esperado" ao dividendo contaria o mesmo dinheiro duas vezes.
"""
from pathlib import Path

import yaml

from .. import armazem

RAIZ = Path(__file__).resolve().parents[2]
MESES_ANO = 12


def carrega_premissas():
    """YAML como base, sobreposto pelo que foi salvo na tela.

    O YAML continua sendo a documentação das premissas - é lá que estão
    os comentários explicando cada uma. A tela grava só o que muda, à
    parte (Supabase ou dados/premissas.json), para que reescrever o
    arquivo não apague os textos.
    """
    with open(RAIZ / "config" / "projecao.yaml", encoding="utf-8") as f:
        base = yaml.safe_load(f) or {}
    try:
        salvas = armazem.premissas_salvas()
        if salvas:
            base = aplica(base, salvas)
    except Exception:
        pass                          # overlay ruim não derruba a tela
    return base


def aplica(premissas, mudancas):
    """Mescla mudanças (aninhadas) sobre as premissas, sem alterar a base."""
    saida = {k: (dict(v) if isinstance(v, dict) else v)
             for k, v in (premissas or {}).items()}
    for chave, valor in (mudancas or {}).items():
        if isinstance(valor, dict) and isinstance(saida.get(chave), dict):
            saida[chave].update({k: v for k, v in valor.items() if v is not None})
        elif valor is not None:
            saida[chave] = valor
    return saida


def salva_premissas(mudancas):
    armazem.salva_premissas(mudancas)


def limpa_premissas():
    """Volta para o que está no YAML."""
    armazem.limpa_premissas()


def tem_premissas_salvas():
    return armazem.tem_premissas_salvas()


def _mensal(taxa_anual_pct):
    """Converte % ao ano em fator mensal composto."""
    return (1 + taxa_anual_pct / 100) ** (1 / MESES_ANO) - 1


def _real(nominal_pct, ipca_pct):
    """Taxa real pela fórmula de Fisher."""
    return ((1 + nominal_pct / 100) / (1 + ipca_pct / 100) - 1) * 100


def perfil_dos_pilares(ativos, renda_fixa, macro_lp, ipca):
    """Estado inicial de cada pilar: quanto vale e como rende.

    Renda variável: valor da posição e o DY médio ponderado, tirado do
    que os ativos de fato pagaram. Caixa: cada título com a taxa real
    que foi contratada, não uma média.
    """
    pilares = {}

    for a in ativos:
        pos = a.get("posicao")
        pilar = a.get("pilar")
        if not (pos and pilar):
            continue
        valor = pos["atual"]
        # DY efetivo do ativo; ETF de índice não distribui, fica em zero
        dy = 0.0
        for m in a.get("metricas", []):
            if m["chave"] == "dy" and m.get("valor"):
                dy = m["valor"]
                break
        p = pilares.setdefault(pilar, {"valor": 0.0, "dy_ponderado": 0.0,
                                       "titulos": []})
        p["valor"] += valor
        p["dy_ponderado"] += valor * dy

    for chave, p in pilares.items():
        p["dy"] = p["dy_ponderado"] / p["valor"] if p["valor"] else 0.0

    # caixa: cada título rende pela própria taxa contratada
    caixa = pilares.setdefault("caixa", {"valor": 0.0, "dy": 0.0, "titulos": []})
    cdi = macro_lp.get("cdi")
    for r in renda_fixa:
        aplicado = float(r.get("valor_aplicado") or 0)
        if not aplicado:
            continue
        taxa = r.get("taxa")
        if taxa is None:
            nominal = cdi                      # sem taxa informada, assume CDI
        elif r["tipo"] == "cdi":
            nominal = cdi * float(taxa) / 100
        elif r["tipo"] == "ipca":
            nominal = ((1 + ipca / 100) * (1 + float(taxa) / 100) - 1) * 100
        else:
            nominal = float(taxa)
        caixa["valor"] += aplicado
        caixa["titulos"].append({"nome": r.get("nome", "—"), "valor": aplicado,
                                 "real": _real(nominal, ipca),
                                 "nominal": nominal})

    if caixa["titulos"]:
        # taxa real média do caixa, ponderada pelo valor de cada título
        total = sum(t["valor"] for t in caixa["titulos"])
        caixa["taxa_real"] = sum(t["valor"] * t["real"] for t in caixa["titulos"]) / total
    else:
        caixa["taxa_real"] = 0.0

    return pilares


def _destino_do_aporte(saldos, alvos, modo):
    """Divide o aporte do mês entre os pilares.

    'rebalancear' manda tudo para o pilar mais distante da meta — é o
    rebalanceamento por aporte, que corrige a alocação sem vender nada
    e sem gerar imposto.
    """
    total = sum(saldos.values())
    if not total or not alvos:
        return {k: 1.0 / len(saldos) for k in saldos} if saldos else {}

    if modo == "proporcional":
        soma_alvos = sum(alvos.values()) or 1
        return {k: alvos.get(k, 0) / soma_alvos for k in saldos}

    faltas = {k: max(0.0, total * alvos.get(k, 0) / 100 - v)
              for k, v in saldos.items()}
    soma = sum(faltas.values())
    if soma <= 0:                       # já está na meta: divide pela meta
        soma_alvos = sum(alvos.values()) or 1
        return {k: alvos.get(k, 0) / soma_alvos for k in saldos}
    return {k: v / soma for k, v in faltas.items()}


def simula(pilares, premissas, alvos, ajuste_pp=0.0):
    """Roda a simulação mensal. Devolve as séries e os totais.

    ajuste_pp desloca a valorização real de todos os pilares — é o que
    gera os cenários pessimista e otimista.
    """
    meses = int(premissas.get("horizonte_anos", 20)) * MESES_ANO
    aporte = float(premissas.get("aporte_mensal", 0))
    reinveste = bool(premissas.get("reinvestir_proventos", True))
    modo = premissas.get("distribuicao_aporte", "rebalancear")
    valorizacao = premissas.get("valorizacao_real_anual") or {}

    saldos = {k: p["valor"] for k, p in pilares.items()}
    dy = {k: p.get("dy", 0.0) for k, p in pilares.items()}

    # fator de crescimento mensal real de cada pilar
    fator = {}
    for k in saldos:
        if k == "caixa":
            # o caixa cresce pela taxa contratada, que já é conhecida;
            # o cenário não a desloca
            fator[k] = _mensal(pilares[k].get("taxa_real", 0.0))
        else:
            fator[k] = _mensal(float(valorizacao.get(k, 0)) + ajuste_pp)

    serie_patrimonio, serie_renda, serie_composicao = [], [], []
    aportado = proventos_acum = 0.0
    inicial = sum(saldos.values())

    for mes in range(1, meses + 1):
        # 1. valorização do mês
        for k in saldos:
            saldos[k] *= 1 + fator[k]

        # 2. proventos: só a renda variável distribui
        renda_mes = sum(saldos[k] * dy[k] / 100 / MESES_ANO
                        for k in saldos if k != "caixa")
        proventos_acum += renda_mes

        # 3. aporte do mês (+ proventos, se reinvestidos)
        entrada = aporte + (renda_mes if reinveste else 0.0)
        aportado += aporte
        pesos = _destino_do_aporte(saldos, alvos, modo)
        for k, w in pesos.items():
            saldos[k] += entrada * w

        if mes % 3 == 0 or mes == meses:        # um ponto por trimestre
            total = sum(saldos.values())
            serie_patrimonio.append((mes / MESES_ANO, total))
            serie_renda.append((mes / MESES_ANO, renda_mes))
            serie_composicao.append({
                "anos": mes / MESES_ANO,
                "inicial": inicial,
                "aportado": aportado,
                "proventos": proventos_acum if reinveste else 0.0,
                "total": total,
            })

    total = sum(saldos.values())
    return {
        "patrimonio": serie_patrimonio,
        "renda": serie_renda,
        "composicao": serie_composicao,
        "final": total,
        "saldos_finais": dict(saldos),
        "inicial": inicial,
        "aportado": aportado,
        "proventos": proventos_acum,
        # o que sobra depois de tirar o que entrou é ganho de capital
        "valorizacao": total - inicial - aportado - (proventos_acum if reinveste else 0),
        "renda_mensal_final": serie_renda[-1][1] if serie_renda else 0.0,
        "renda_mensal_hoje": sum(pilares[k]["valor"] * dy[k] / 100 / MESES_ANO
                                 for k in pilares if k != "caixa"),
    }


def projeta(dados, premissas):
    """Monta o cenário base e a banda pessimista/otimista."""
    macro = dados["macro"]
    lp = premissas.get("macro_longo_prazo") or {}
    cdi = lp.get("cdi") if lp.get("cdi") is not None else macro.get("cdi")
    ipca = lp.get("ipca") if lp.get("ipca") is not None else macro.get("ipca_12m")
    macro_lp = {"cdi": cdi, "ipca": ipca}

    alvos = dados.get("alocacao_alvo") or {}
    pilares = perfil_dos_pilares(dados["ativos"], dados["renda_fixa"],
                                 macro_lp, ipca)

    cen = premissas.get("cenarios") or {}
    base = simula(pilares, premissas, alvos, 0.0)
    pess = simula(pilares, premissas, alvos, float(cen.get("pessimista", -3)))
    otim = simula(pilares, premissas, alvos, float(cen.get("otimista", 3)))

    # retorno total de cada pilar = valorização premissa + o que ele paga
    # em proventos; serve para o leitor conferir se a premissa é plausível
    valorizacao = premissas.get("valorizacao_real_anual") or {}
    resumo_pilares = []
    for chave, p in pilares.items():
        if chave == "caixa":
            resumo_pilares.append({
                "chave": chave, "valor": p["valor"], "dy": None,
                "valorizacao": None, "total": p.get("taxa_real", 0.0),
                "titulos": p.get("titulos", []),
            })
        else:
            v = float(valorizacao.get(chave, 0))
            # o dividendo acompanha a inflação, então em moeda de hoje o
            # DY nominal já é o rendimento real
            resumo_pilares.append({
                "chave": chave, "valor": p["valor"], "dy": p["dy"],
                "valorizacao": v, "total": v + p["dy"], "titulos": [],
            })

    return {
        "base": base, "pessimista": pess, "otimista": otim,
        "pilares": resumo_pilares,
        "macro_lp": macro_lp,
        "usando_macro_de_hoje": lp.get("cdi") is None,
        "premissas": premissas,
    }
